[LTC] Upstream several internal ops

desertfire opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #69815
* __->__ #69716
* #69621

Summary: To prepare for the landing of LazyTensor and LazyGraphExecutor,
- arithmetic_ir_ops.h
- cast.h
- device_data.h
- expand.h
- generic.h
- scalar.h

Differential Revision: [D32999410](https://our.internmc.facebook.com/intern/diff/D32999410)

desertfire committed:
[LTC] Upstream several internal ops

Summary: To prepare for the landing of LazyTensor and LazyGraphExecutor,
- arithmetic_ir_ops.h
- cast.h
- device_data.h
- expand.h
- generic.h
- scalar.h

[ghstack-poisoned]

desertfire commented:
@desertfire has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32999410).

desertfire committed:
Update on "[LTC] Upstream several internal ops"


Summary: To prepare for the landing of LazyTensor and LazyGraphExecutor,
- arithmetic_ir_ops.h
- cast.h
- device_data.h
- expand.h
- generic.h
- scalar.h

Differential Revision: [D32999410](https://our.internmc.facebook.com/intern/diff/D32999410)

[ghstack-poisoned]

desertfire commented:
@desertfire has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32999410).

wconstab commented on torch/csrc/lazy/core/internal_ops/arithmetic_ir_ops.cpp:
```diff
@@ -0,0 +1,45 @@
+namespace torch {
+namespace lazy {
+
+NodePtr operator+(const Value& node1, const Value& node2) {
```
these ops in particular are ones i think we don't want to put in core.  keeping them in ts_backend folder for is fine, but hopefully we'll be able to delete them soon once we codegen mul and a few other ops.

wconstab reviewed (approved):
would prefer to locate the ops under ts_backend but otherwise lgtm.

desertfire committed:
Update on "[LTC] Upstream several internal ops"


Summary: To prepare for the landing of LazyTensor and LazyGraphExecutor,
- arithmetic_ir_ops.h
- cast.h
- device_data.h
- expand.h
- generic.h
- scalar.h

Differential Revision: [D32999410](https://our.internmc.facebook.com/intern/diff/D32999410)

[ghstack-poisoned]

desertfire commented:
> would prefer to locate the ops under ts_backend but otherwise lgtm.

Moved based on the suggestion.

desertfire committed:
Update on "[LTC] Upstream several internal ops"


Summary: To prepare for the landing of LazyTensor and LazyGraphExecutor,
- arithmetic_ir_ops.h
- cast.h
- device_data.h
- expand.h
- generic.h
- scalar.h

Differential Revision: [D32999410](https://our.internmc.facebook.com/intern/diff/D32999410)

[ghstack-poisoned]

desertfire commented:
@desertfire has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32999410).

desertfire committed:
Update on "[LTC] Upstream several internal ops"


Summary: To prepare for the landing of LazyTensor and LazyGraphExecutor,
- arithmetic_ir_ops.h
- cast.h
- device_data.h
- expand.h
- generic.h
- scalar.h

Differential Revision: [D32999410](https://our.internmc.facebook.com/intern/diff/D32999410)

[ghstack-poisoned]

desertfire commented:
@desertfire has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32999410).

desertfire committed:
Update on "[LTC] Upstream several internal ops"


Summary: To prepare for the landing of LazyTensor and LazyGraphExecutor,
- arithmetic_ir_ops.h
- cast.h
- device_data.h
- expand.h
- generic.h
- scalar.h

Differential Revision: [D32999410](https://our.internmc.facebook.com/intern/diff/D32999410)

[ghstack-poisoned]

desertfire committed:
Update on "[LTC] Upstream several internal ops"


Summary: To prepare for the landing of LazyTensor and LazyGraphExecutor,
- arithmetic_ir_ops.h
- cast.h
- device_data.h
- expand.h
- generic.h
- scalar.h

Differential Revision: [D32999410](https://our.internmc.facebook.com/intern/diff/D32999410)

[ghstack-poisoned]

desertfire committed:
Update on "[LTC] Upstream several internal ops"


Summary: To prepare for the landing of LazyTensor and LazyGraphExecutor,
- arithmetic_ir_ops.h
- cast.h
- device_data.h
- expand.h
- generic.h
- scalar.h

Differential Revision: [D32999410](https://our.internmc.facebook.com/intern/diff/D32999410)

[ghstack-poisoned]

desertfire committed:
Update on "[LTC] Upstream several internal ops"


Summary: To prepare for the landing of LazyTensor and LazyGraphExecutor,
- arithmetic_ir_ops.h
- cast.h
- device_data.h
- expand.h
- generic.h
- scalar.h

Differential Revision: [D32999410](https://our.internmc.facebook.com/intern/diff/D32999410)

[ghstack-poisoned]
